File: A2Z_sheet/Stack_and_Queue/Learning/07_check_for_balanced_parentheses/solution.ts
function isValid(s: string): boolean {
    const stack: string[] = [];
    
    for (const char of s) {
        if (char === '(' || char === '{' || char === '[') {
            stack.push(char);
            continue;
        }

        const top: string | undefined = stack.pop();
        
        if ((char === ')' && top === '(') || 
            (char === '}' && top === '{') || 
            (char === ']' && top === '[')) {
            continue;
        } else {
            return false;
        }
    }
    
    return stack.length === 0;
}

File: A2Z_sheet/Stack_and_Queue/Learning/07_check_for_balanced_parentheses/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
    
        for char in s.chars() {
            match char {
                '(' | '{' | '[' => stack.push(char),
                ')' => if stack.pop() != Some('(') { return false; },
                '}' => if stack.pop() != Some('{') { return false; },
                ']' => if stack.pop() != Some('[') { return false; },
                _ => return false,
            }
        }
        
        stack.is_empty()
    }
}

